OK-41 ModelInstantiator

pcrist opened the pull request.

pcrist commented:
ModelInstantiator needs to implement validation.

pcrist commented on src/Models/ModelInstantiator.php:
```diff
@@ -0,0 +1,120 @@
+        object $data_object
+    ): array
+    {
+        return (array) $data_object;
```
This assumes the data object is hydrated. Can we assume that or should we iterate over the object properties to make sure all of them are included?

pcrist commented on src/Models/ModelInstantiator.php:
```diff
@@ -0,0 +1,120 @@
+     */
+    public function dataObjectToDatabaseModel(
+        object $data_object
+    ): array
```
The spec indicated that `string $class` should be a parameter here, but I think we can get that value from `$data_object::class`.

bbotto-pdga commented on src/Models/ModelInstantiator.php:
```diff
@@ -0,0 +1,120 @@
+            $property_reflection = new ReflectionProperty($data_object::class, $property);
+
+            // Get the Column attribute.
+            $attribute = $property_reflection->getAttributes(Column::class);
```
I'm not sure if there will ever be a case where we call this method with a Data Object that has a property that doesn't map to a Column.  If so, this `$attribute` array would be empty and the next line would crash.  Can the next line be skipped for properties that are not Column-attributed?

bbotto-pdga commented on src/Models/ModelInstantiator.php:
```diff
@@ -0,0 +1,120 @@
+        foreach (array_keys(get_class_vars($class)) as $property)
+        {
+            $property_reflection = new ReflectionProperty($class, $property);
+            $attribute           = $property_reflection->getAttributes(Column::class);
```
Same note about properties that are not Column-attributed.

bbotto-pdga reviewed (approved):
Awesome.  We're slowly taking over the world.

drew-acheson commented on src/Models/ModelInstantiator.php:
```diff
@@ -0,0 +1,120 @@
+class ModelInstantiator
+{
+    /**
+     * Converts an associate array to a validated instance of a data object class.
```
Poor array hasn't made full partner yet.

pcrist replied:
👀

pcrist commented:
I updated ModelInstantiator to be more robust and specific about properties having Column attributes.

I also replaced the placeholder class for tests with a dedicated class that I think we can keep. Hopefully this is less confusing.

bbotto-pdga commented on src/Models/ModelInstantiator.php:
```diff
@@ -0,0 +1,143 @@
+        // Assign all public properties.
+        foreach ($this->dataObjectPropertyColumns($class) as $property => $column)
+        {
+            $instance->{$property} = $arr[$property] ?? null;
```
These need to remain unset if the property doesn't exist on the array, they can't be set to null.  We need to be able to update records partially, e.g. just the `FirstName` of a member.  Let's say an update request comes in with this array:

```
['firstName' => 'Ben']
```

We map that to a `ModelInstantiatorTestObject`.  This code would set `LastName` etc. to null.  Then, down the line, the Repository code would set `LastName` etc. to NULL, which isn't right.  We want `LastName` etc. to be ignored.

bbotto-pdga commented on src/Models/ModelInstantiator.php:
```diff
@@ -0,0 +1,143 @@
+        foreach ($this->dataObjectPropertyColumns($data_object::class) as $property => $column)
+        {
+            // Assign key based on the name property of the Column attribute.
+            $model_array[$column] = $data_object->{$property};
```
Similar comment here as the one above: This needs to check if `property` is set, hopefully using the helper methods that Drew wrote in the `ValidationEnforcer`.  If `$property` is uninitialized, this will result in a warning or error depending on PHP config.

pcrist replied:
Thanks. I added a basic check for the time being, and I'll plan to update once the validation code is merged.
